Fix docs directory structure to reflect diátaxis

davidekete opened the pull request:
## Description


## Checklist

- [ x ] Runs `make check` successfully.
- [ ] Retains code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

rkratky commented:
@davidekete, thanks for submitting this! To fix the spellcheck and one cross-doc link, I submitted a PR against your branch: https://github.com/davidekete/netplan/pull/2

davidekete commented:
> @davidekete, thanks for submitting this! To fix the spellcheck and one cross-doc link, I submitted a PR against your branch: [davidekete#3](https://github.com/davidekete/netplan/pull/3)

Thank you. I just merged the PR.

rkratky commented:
Two notes:

* @davidekete, as discussed on Matrix, could you please rename the main/landing files in the individual categories to `index.md`? This naming convention is used in all other Ubuntu docs published on RTD, so it'd be good to keep it consistent. Also, it'd avoid the duplication that results from Sphinx building the landing page in two places when the index file is absent.
* @slyon, while this change in directory structure is certainly an improvement for the repo, it'll mean some existing URLs would no longer work. Therefore, we need redirects. My question is: do you think it would be enough to add redirects for the main categories only (i.e. explanation, howto, ...), or do we need redirects for all existing pages?

slyon commented:
> * do we need redirects for all existing pages?

I think the most important one would be the `netplan-yaml` reference. Maybe that in combination with the main categories. And also making sure that our website still points to the correct reference, e.g.: https://netplan.io/reference

davidekete commented:
> Two notes:
> 
> * @davidekete, as discussed on Matrix, could you please rename the main/landing files in the individual categories to `index.md`? This naming convention is used in all other Ubuntu docs published on RTD, so it'd be good to keep it consistent. Also, it'd avoid the duplication that results from Sphinx building the landing page in two places when the index file is absent.
> * @slyon, while this change in directory structure is certainly an improvement for the repo, it'll mean some existing URLs would no longer work. Therefore, we need redirects. My question is: do you think it would be enough to add redirects for the main categories only (i.e. explanation, howto, ...), or do we need redirects for all existing pages?

@rkratky, I have renamed them as `index.md` and updated the references to match the names.

rkratky commented:
@slyon, I believe this is good to be merged. Could you please have a look?

slyon commented on doc/conf.py:
```diff
@@ -121,7 +121,7 @@
 redirects = {
     'README.md': '/',
-    'netplan': '/netplan-yaml',
+    'netplan': '/reference/netplan-yaml',
```
todo: I'm still missing a redirect for the old `netplan-yaml` here.

In the end, we need to be able to redirect something like https://netplan.io/reference#routing to https://netplan.readthedocs.io/en/latest/reference/netplan-yaml/#routing

I was unable to make the `sphinx-reredirects` redirect any anchor (i.e. the URI fragment after the hashtag `#routing`), but this is important for the `netplan-yaml` page, as there are lots of deep links aroud the internet pointing to specific configuration and we would break all of them if we don't support redirecting to the proper anchor.

@rkratky Do you have any idea how anchor redirects could be solved?

davidekete replied:
When you say you're missing a redirect for the old `netplan-yaml`, what do you mean @slyon?

slyon commented on doc/howto/multi-nic-vm-host-with-bonds-and-vlans.md:
todo: I see plenty of `CRITICAL: Directive "include": file not found:` from this file. Probably needs a `../reuse/...`

davidekete replied:
Resolved

slyon commented on doc/howto/single-nic-vm-host-with-vlans.md:
todo: I see plenty of `CRITICAL: Directive "include": file not found:` from this file. Probably needs a `../reuse/...`

davidekete replied:
resolved

slyon commented on doc/howto/single-nic-vm-host.md:
todo: I see plenty of `CRITICAL: Directive "include": file not found:` from this file. Probably needs a `../reuse/...`

davidekete replied:
resolved

slyon commented on doc/howto/netplan-everywhere.md:
todo: We need to update links to `netplan-everywhere` afterwards, especially here: https://ubuntu.com/tutorials/access-desktop-networkmanager-settings-through-netplan#6-enjoy-netplan-on-desktop

davidekete replied:
Can you provide a little more context to this? Does this have to be done here, or do we need to update it directly in the linked docs? I'm assuming it's the latter.

davidekete replied:
Oh I see. Thanks for the help. I have added the redirect.

slyon replied:
Thank you for adding a `netplan-everyhwere` redirect, I think that's the correct thing to do!
Unfortunately, it doesn't seem to be working as expected. When I try to trigger the redirect on the PR branch of readthe docs (https://netplan--520.org.readthedocs.build/en/520/netplan-everywhere/), I get redirected to https://netplan--520.org.readthedocs.build/howto/netplan-everywhere which returns a "404".

Similarly, on the main docs page at readthedocs, if I enter https://netplan.readthedocs.io/netplan-everywhere/ it returns 404/"Page not found", I guess that's due to to missing prefix `en/stable/`, `en/latest/`, `en/520/`, ... So it seems like the absolute redirect to `/howto/...` isn't working properly, and we need to make it a relative redirect in some way.

slyon replied:
@davidekete Looks like I've been able to fix those redirects by using relative top-level redirects (i.e. `../` instead of `/`). I pushed commit https://github.com/canonical/netplan/pull/520/commits/c5c732b0207883a044861d8e2cea41efb469a54c into your branch to fix that.

davidekete replied:
Thanks for your help @slyon! I appreciate it.

slyon reviewed (changes requested):
Sorry for the late review and thank you very much for your contribution to Netplan!

In general this looks fine, but it also breaks lots of references around the internet. I wonder if the benefit is great enough to justify that? I see that have this new structure would improve the maintainability going forward... @rkratky what's your stance on this? Is there any recommendation from the technical authors?

I'd be fine with merging it once we fix the broken internal references (see inline comments of certian "howto" files) and the outside-references to `netplan-yaml`, using anchor redirects. This would be the bare minimum, I guess.

rkratky commented:
> In general this looks fine, but it also breaks lots of references around the internet.

Yes, that's my main concern, too.

> I wonder if the benefit is great enough to justify that? I see that have this new structure would improve the maintainability going forward... @rkratky what's your stance on this? Is there any recommendation from the technical authors?

I believe the benefits outweigh the negatives (I suggested this change).

slyon commented:
> I believe the benefits outweigh the negatives (I suggested this change).

ACK, fair enough. Would you be able to assist with the critical redirects for the `netplan-yaml` page, then? E.g. https://netplan.io/reference#routing to https://netplan.readthedocs.io/en/latest/reference/netplan-yaml/#routing ?

rkratky commented:
> Would you be able to assist with the critical redirects

Yes, I hope. Just acknowledging now because I'm not sure if I'll get to it this week, i.e. before the Sprints.

davidekete commented:
> > Would you be able to assist with the critical redirects
> 
> Yes, I hope. Just acknowledging now because I'm not sure if I'll get to it this week, i.e. before the Sprints.

Hi @rkratky, @slyon, I could work on this myself to speed up the process. I just need a little more context to what needs to be done and i'll get righ on it.

davidekete commented:
> > > > Would you be able to assist with the critical redirects
> > > 
> > > 
> > > Yes, I hope. Just acknowledging now because I'm not sure if I'll get to it this week, i.e. before the Sprints.
> > 
> > 
> > Hi @rkratky, @slyon, I could work on this myself to speed up the process. I just need a little more context to what needs to be done and i'll get righ on it.
> 
> @davidekete Thanks, that's appreciated! See my inline comments above, there are some easy TODOs in there for you, e.g. fixing the include path for the "reuse" documents.
> 
> After that we need to find a way to allow Sphinx/ReadTheDocs to redirect URLs with an anchor, like this: https://netplan.io/reference#routing to https://netplan.readthedocs.io/en/latest/reference/netplan-yaml/#routing (preserving the "#routing" fragment). I don't know how this works and wasn't able to find a way when quickly looking into that.
> 
> Furthermore, there seems to be a merge conflict in `doc/netplan-yaml.md` by now (sorry for taking so long), which needs to be resolved.

Hi @slyon, I fixed most of the Todos. I tried to find a way to allow Sphinx/ReadTheDocs to redirect URLs with an anchor, but I couldn't; this is new territory for me. Could you explain the background of the problem? and If you figure out a way to solve it, could you explain that to me too?

slyon commented on doc/reference/netplan-yaml.md:
```diff
@@ -2192,4 +2192,4 @@ consumer of that back end. Currently, this is only used with `NetworkManager`.
     > Can be used as a fallback mechanism to missing key-file settings.
 
-<!--- vim: ft=markdown -->
+<!--- vim: ft=markdown -->
```
nitpick (non-blocking): This change seems to be unrelated and we should consider dropping it.

```diff
-<!--- vim: ft=markdown -->
+<!--- vim: ft=markdown -->
\ No newline at end of file
```

davidekete replied:
Hmmm, I didn't add this myself (Tehnically, I did, to resolve the conflict). It's part of the current netplan-yaml file on the main branch.

slyon reviewed (changes requested):
> Hi @slyon, I fixed most of the Todos. I tried to find a way to allow Sphinx/ReadTheDocs to redirect URLs with an anchor, but I couldn't; this is new territory for me. Could you explain the background of the problem? and If you figure out a way to solve it, could you explain that to me too?

Thanks for solving the merge conflicts and the `../reuse/...` imports! I left two more inline-comments, about an (apparently?) unrelated change and an additional redirect.

Regarding the anchor redirect, that you asked about:
We need to find a way to redirect URLs with an anchor, like this: https://netplan.io/reference#routing or https://netplan.readthedocs.io/en/latest/netplan-yaml/#routing to the new URL at https://netplan.readthedocs.io/en/latest/reference/netplan-yaml/#routing – The important part being the `#routing` fragment/section here.

There are lots of links out on AskUbuntu, Launchpad and other places, that are deep links into the reference document (`netplan-yaml`) and we need to make sure those are still working after this PR is merged.

I don't know how this works either and wasn't able to find a way when quickly looking into that. I hope for @rkratky experience here.

rkratky commented:
> I don't know how this works either and wasn't able to find a way when quickly looking into that. I hope for @rkratky experience here.

Just dropping in to mention I'm not ignoring this PR/discussion -- I just couldn't find the time for investigating these two weeks. I hope to get back to normal next week.

davidekete commented:
Hi guys, just bumping this issue. I tried my hand at it and posted the issue on [StackOverflow](https://stackoverflow.com/questions/79263377/how-do-i-maintain-url-fragments-when-using-sphinx-and-sphinx-reredirects-to-hand) for help. Have you guys had any luck on your end? @rkratky @slyon

davidekete commented:
Hi @slyon @rkratky bump

rkratky commented:
@davidekete, @slyon 

I submitted #567, which implements these changes -- it was simpler to start a new PR to avoid conflicts after such a long time has passed.

Apologies for taking so criminally long with responding to this and finalizing it. All credit for the actual work goes to @davidekete. In the new PR, I only added internal redirects to handle all of the changes and added some minor updates.

I really don't have a good excuse for neglecting this for so long, and I'm truly sorry. Somehow, I just kept postponing the work, and then it got totally out of hand. Something always came up that seemed more pressing at the time. Again, my apologies.

......

I believe the new PR is good to be merged. I tested the redirects as I could, and I didn't find any case where it wouldn't work as expected.
